src/lib/OccAnalysis.hs

```diff
@@ -0,0 +1,559 @@
+    RunIO bd ->
+      -- I think IxAll is right, since indexing with a value of Unit type
+      -- does mean indexing with every possible value of the index type.
+      RunIO <$> oneShot a IxAll bd
+    RunInit _ ->
+      -- Though this is probably not too hard to implement.  Presumably
+      -- the lambda is one-shot.
+      error "Expecting to do occurrence analysis before lowering."
+    CatchException _ ->
+      error "Expecting to do occurrence analysis after exception resolution."
+      -- But it's one-shot if we need it: CatchException <$> oneShot bd
+    Linearize _ -> error "Expecting to do occurrence analysis after AD."
+    Transpose _ -> error "Expecting to do occurrence analysis after AD."
+    Seq _ _ _ _ ->
+      error "Expecting to do occurrence analysis before destination passing."
+    RememberDest _ _ ->
+      error "Expecting to do occurrence analysis before destination passing."
+
+-- Occurrence analysis of the body of a one-shot unary lambda Atom.
+oneShot :: Access n -> IxExpr n -> Atom n -> OCCM n (Atom n)
+oneShot acc binderSummary bd =
+  occLam bd \b' body -> extend b' (sink binderSummary) do
+    occ (sink acc) body
+
+-- Occurrence analysis of the body of a one-shot biary lambda Atom.
+oneShot2 :: Access n -> IxExpr n -> IxExpr n -> Atom n -> OCCM n (Atom n)
+oneShot2 acc binderSummary1 binderSummary2 bd =
+  occLam bd \b' (AtomicBlock body)
+    -> AtomicBlock <$> do
+      extend b' (sink binderSummary1) do
+        oneShot (sink acc) (sink binderSummary2) body
+
+-- Going under a lambda binder.
+occLam :: Atom n
+       -> (forall l. DExt n l
+           => (NameBinder 'AtomNameC n l) -> Block l -> OCCM l (Block l))
+       -> OCCM n (Atom n)
+occLam bd cont = case bd of
+  Lam (LamExpr (LamBinder b ty arr eff) body) -> do
+    ty' <- occ accessOnce ty  -- TODO Is free variable counting enough here?
+    (Abs b' (body' `PairE` eff')) <-
+      refreshAbs (Abs (b:>ty') (body `PairE` eff))
+        \(b':>_) (body' `PairE` eff') -> do
+          countFreeVarsAsOccurrences eff'
+          body'' <- cont b' body'
+          return $ Abs b' (body'' `PairE` eff')
+    return $ Lam $ LamExpr (LamBinder b' ty' arr eff') body'
+  _ -> error "Not a lambda expression"
+
+instance HasOCC Atom where
+  occ a atom = case atom of
+    Lam _ -> error "Expecting all Lam Atoms to be covered by their contexts"
+    ACase _ _ _ -> error "Unexpected ACase outside of Simplify"
+    _ -> generic
+    where
+      generic = toE <$> (occ a $ fromE atom)
+  {-# INLINE occ #-}
+
+instance HasOCC TabLamExpr where
+  occ _ view = inlinedLater view
+  {-# INLINE occ #-}
+
+instance HasOCC (ComposeE PrimOp Atom) where
+  occ _ expr@(ComposeE (PrimEffect _ (MExtend (BaseMonoid _ combine) val))) = do
+    valIx <- summary val
+    -- Treat the combining function as inlined here and called once
+    void $ oneShot2 accessOnce (Deterministic []) valIx combine
+    -- Do not change it
```

Comment:
Is that correct? What if the combining function has some read code that was eliminated, and that was reflected in the accesses?

Reply:
Hm.  That sounds like a potential bug.  Noted in a comment.  Let me know if you think we should fix it now.